orrery/include/orrery/component.h

```diff
@@ -0,0 +1,69 @@
+// Concord
+//
+// Copyright (c) 2021 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the
+// "License").  You may not use this product except in compliance with the
+// Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include <utility>
+#include <type_traits>
+
+#include "kvstream.h"
+#include "orrery_msgs.cmf.hpp"
+#include "Logger.hpp"
+#include <iostream>
+
+namespace concord::orrery {
+
+// A component provides an interface to handle `AllMsgs`
+class IComponent {
+ public:
+  virtual ~IComponent(){};
+  virtual void handle(ComponentId from, AllMsgs&&) = 0;
+};
+
+template <typename ComponentImpl>
+class Component : public IComponent {
```

Comment:
I love this pattern - having runtime polymorphic behaviour and not forcing users to have virtual methods and allocate dynamically themselves!